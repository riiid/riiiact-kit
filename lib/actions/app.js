export const TOGGLE_DRAWER = 'TOGGLE_DRAWER';

export const SHOW_SNACKBAR = 'SHOW_SNACKBAR';
export const HIDE_SNACKBAR = 'HIDE_SNACKBAR';

export const SHOW_PROGRESS = 'SHOW_PROGRESS';
export const HIDE_PROGRESS = 'HIDE_PROGRESS';

export const SHOW_TOOLBAR = 'SHOW_TOOLBAR';
export const HIDE_TOOLBAR = 'HIDE_TOOLBAR';

export const FIREBASE_AVAILABLE = 'FIREBASE_AVAILAVLE';

export const toggleDrawer = () => {
  return {
    type: TOGGLE_DRAWER
  };
};

export const showSnackbar = (message, props = {}) => {
  return {
    type: SHOW_SNACKBAR,
    message,
    props
  };
};

export const hideSnackbar = () => {
  return {
    type: HIDE_SNACKBAR
  };
};

export const showProgress = () => {
  return {
    type: SHOW_PROGRESS
  };
};

export const hideProgress = () => {
  return {
    type: HIDE_PROGRESS
  };
};

export const showToolbar = () => {
  return {
    type: SHOW_TOOLBAR
  };
};

export const hideToolbar = () => {
  return {
    type: HIDE_TOOLBAR
  };
};

export const firebaseAvailable = available => {
  return {
    type: FIREBASE_AVAILABLE,
    available
  };
};
