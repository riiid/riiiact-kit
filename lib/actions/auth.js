export const REQUEST_AUTH = 'REQUEST_AUTH';
export const RECEIVE_AUTH_SUCCESS = 'RECIEVE_AUTH_SUCCESS';
export const RECEIVE_AUTH_FAILURE = 'RECIEVE_AUTH_FAILURE';

export const requestAuth = () => ({type: REQUEST_AUTH});

export const receiveAuthSuccess = user => {
  return {
    type: RECEIVE_AUTH_SUCCESS,
    user
  };
};

export const receiveAuthFailure = error => {
  return {
    type: RECEIVE_AUTH_FAILURE,
    error
  };
};
