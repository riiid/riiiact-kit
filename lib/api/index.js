import initFirebase from 'api/firebase';

export const firebase = initFirebase();
