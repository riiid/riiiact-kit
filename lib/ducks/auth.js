export const REQUEST_AUTH = 'REQUEST_AUTH';
export const RECEIVE_AUTH_SUCCESS = 'RECIEVE_AUTH_SUCCESS';
export const RECEIVE_AUTH_FAILURE = 'RECIEVE_AUTH_FAILURE';

export const requestAuth = () => ({type: REQUEST_AUTH});

export const receiveAuthSuccess = user => {
  return {
    type: RECEIVE_AUTH_SUCCESS,
    user
  };
};

export const receiveAuthFailure = error => {
  return {
    type: RECEIVE_AUTH_FAILURE,
    error
  };
};

const INITIAL_STATE = {
  authenticated: false,
  user: null
};

export default (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case RECEIVE_AUTH_SUCCESS:
      return {
        ...state,
        authenticated: true,
        user: action.user,
        error: null
      };
    case RECEIVE_AUTH_FAILURE:
      return {
        ...state,
        authenticated: false,
        user: null,
        error: action.error
      };
    default:
      return state;
  }
};
