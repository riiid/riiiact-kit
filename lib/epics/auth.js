import {Observable} from 'rxjs/Observable';
import 'rxjs/add/observable/merge';
import 'rxjs/add/operator/filter';
import 'rxjs/add/operator/mergeMap';
import 'rxjs/add/operator/startWith';
import 'rxjs/add/operator/catch';
import 'rxjs/add/operator/share';
import 'rxjs/add/operator/share';
import {
  REQUEST_AUTH, receiveAuthSuccess, receiveAuthFailure
} from 'ducks/auth';
import {
  showProgress, hideProgress, showSnackbar, firebaseAvailable
} from 'ducks/app';

export const auth$$ = (action$, store, {firebase}) => {
  const {auth, db} = firebase;
  return action$.ofType(REQUEST_AUTH)
    .mergeMap(() => {
      const auth$ = auth.onAuthStateChanged$().share();
      const signIn$ = auth$
        .filter(user => user === null)
        .mergeMap(() => auth.signInAnonymously$());
      const login$ = auth$
        .filter(user => user !== null)
        .do(user => {
          db.ref('online-users').child(user.uid).set(true);
          db.ref('online-users').child(user.uid).onDisconnect().set(null);
        });
      return Observable.merge(signIn$, login$)
        .mergeMap(({uid, email, displayName, isAnonymouse}) => {
          return [
            hideProgress(),
            receiveAuthSuccess({uid, email, displayName, isAnonymouse})
          ];
        })
        .catch(err => [
          hideProgress(),
          showSnackbar('로그인 중 오류 발생'),
          receiveAuthFailure(err)
        ])
        .startWith(
          showProgress(),
          firebaseAvailable(db !== null && auth !== null)
        );
    });
};
