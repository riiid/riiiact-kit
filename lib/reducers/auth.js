import {RECEIVE_AUTH_SUCCESS, RECEIVE_AUTH_FAILURE} from 'actions/auth';

const INITIAL_STATE = {
  authenticated: false,
  user: null
};

export default (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case RECEIVE_AUTH_SUCCESS:
      return {
        ...state,
        authenticated: true,
        user: action.user,
        error: null
      };
    case RECEIVE_AUTH_FAILURE:
      return {
        ...state,
        authenticated: false,
        user: null,
        error: action.error
      };
    default:
      return state;
  }
};
