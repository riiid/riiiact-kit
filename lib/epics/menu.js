import {Observable} from 'rxjs/Observable';
import keys from 'lodash/keys';
import {firebase} from 'api';
import {
  REQUEST_MENU, receiveMenuSuccess, receiveMenuFailure
} from 'actions/menu';
import {showProgress, hideProgress} from 'actions/app';

const count = () => {
  const {db} = firebase;
  db.ref('stats').child('totalViews').transaction$(current => {
    if (current === null) {
      return 1;
    }
    return current + 1;
  }).subscribe();
};

export const menuEpic = (action$, store) => {
  const {db} = firebase;
  const {dispatch} = store;
  return action$.ofType(REQUEST_MENU)
    .do(() => dispatch(showProgress()))
    .mergeMap(() => {
      count();
      const tv$ = db.ref('stats').child('totalViews').observe$('value')
        .map(snapshot => snapshot.val());
      const cv$ = db.ref('online-users').observe$('value')
        .map(snapshot => snapshot.val());

      return Observable.combineLatest(tv$, cv$, (tv, cv) => {
        return {
          totalViews: tv,
          currentViews: keys(cv).length
        };
      })
      .do(() => dispatch(hideProgress()))
      .map(receiveMenuSuccess)
      .catch(err => {
        dispatch(hideProgress());
        return Observable.of(receiveMenuFailure(err));
      });
    });
};
