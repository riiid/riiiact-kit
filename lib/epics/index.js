import {combineEpics} from 'redux-observable';
import {authEpic} from 'epics/auth';
import {menuEpic} from 'epics/menu';

export default combineEpics(
  authEpic,
  menuEpic
);
