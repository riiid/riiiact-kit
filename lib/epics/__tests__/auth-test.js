jest.mock('api');
jest.mock('store');

import Rx from 'rxjs';
import createStore from 'store';
import {requestAuth } from 'actions/auth';
import {createEpicMiddleware} from 'redux-observable';
import {authEpic} from 'epics/auth';
import {firebase} from 'api';

describe('epics/auth', () => {
  let store;

  beforeEach(() => store = createStore([createEpicMiddleware(authEpic)]));

  it('when requestAuth success', () => {
    firebase.auth.onAuthStateChanged$
    .mockImplementationOnce(() => {
      return Rx.Observable.of({uid: 'UID', isAnonymouse: true});
    });

    store.dispatch(requestAuth());
    expect(store.getActions()).toMatchSnapshot();
  });

  it('should call signInAnonymousely when requestAuth with null user', () => {
    firebase.auth.onAuthStateChanged$
    .mockImplementationOnce(() => {
      return Rx.Observable.of(null);
    });

    store.dispatch(requestAuth());
    expect(store.getActions()).toMatchSnapshot();
    expect(firebase.auth.signInAnonymously).toHaveBeenCalled();
  });

  it('when requestAuth fail', () => {
    firebase.auth.onAuthStateChanged$
    .mockImplementationOnce(() => {
      return Rx.Observable.throw(new Error('ERROR'));
    });

    store.dispatch(requestAuth());
    expect(store.getActions()).toMatchSnapshot();
  });
});
