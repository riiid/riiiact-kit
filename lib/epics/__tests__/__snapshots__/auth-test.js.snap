// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`epics/auth should call signInAnonymousely when requestAuth with null user 1`] = `
Array [
  Object {
    "type": "REQUEST_AUTH",
  },
  Object {
    "type": "SHOW_PROGRESS",
  },
  Object {
    "available": true,
    "type": "FIREBASE_AVAILAVLE",
  },
  Object {
    "type": "HIDE_PROGRESS",
  },
  Object {
    "type": "RECIEVE_AUTH_SUCCESS",
    "user": Object {
      "displayName": null,
      "email": null,
      "isAnonymouse": true,
      "uid": "UID",
    },
  },
]
`;

exports[`epics/auth when requestAuth fail 1`] = `
Array [
  Object {
    "type": "REQUEST_AUTH",
  },
  Object {
    "type": "SHOW_PROGRESS",
  },
  Object {
    "available": true,
    "type": "FIREBASE_AVAILAVLE",
  },
  Object {
    "type": "HIDE_PROGRESS",
  },
  Object {
    "message": "로그인 중 오류 발생",
    "props": Object {},
    "type": "SHOW_SNACKBAR",
  },
  Object {
    "error": [Error: ERROR],
    "type": "RECIEVE_AUTH_FAILURE",
  },
]
`;

exports[`epics/auth when requestAuth success 1`] = `
Array [
  Object {
    "type": "REQUEST_AUTH",
  },
  Object {
    "type": "SHOW_PROGRESS",
  },
  Object {
    "available": true,
    "type": "FIREBASE_AVAILAVLE",
  },
  Object {
    "type": "HIDE_PROGRESS",
  },
  Object {
    "type": "RECIEVE_AUTH_SUCCESS",
    "user": Object {
      "displayName": null,
      "email": null,
      "isAnonymouse": true,
      "uid": "UID",
    },
  },
]
`;
